### ISSUE TITLE
roberta preprocessing help

### ISSUE THREAD

**Issue by rush86999** [pytorch/fairseq#994]
hi,

so I 'm trying to extract features in the preprocessing step of my model and I might be doing this wrong because I 'm not sure if roberta can do this or not. 

so here's the error:
```
Traceback (most recent call last):
  File "paraphrase_train_roberta.py", line 308, in <module>
    train(args, device_id)
  File "paraphrase_train_roberta.py", line 145, in train
    trainer.train(train_iter_fct, args.train_steps)
  File "/home/ec2-user/snips-extractor/src/models/roberta_trainer.py", line 152, in train
    for i, batch in enumerate(train_iter):
  File "/home/ec2-user/snips-extractor/src/models/data_loader_roberta.py", line 154, in __iter__
    for batch in self.cur_iter:
  File "/home/ec2-user/snips-extractor/src/models/data_loader_roberta.py", line 249, in __iter__
    batch = Batch(minibatch, self.device, self.is_test)
  File "/home/ec2-user/snips-extractor/src/models/data_loader_roberta.py", line 37, in __init__
    src = roberta.extract_features(pre_src)
  File "/home/ec2-user/.cache/torch/hub/pytorch_fairseq_master/fairseq/models/roberta/hub_interface.py", line 87, in extract_features
    return_all_hiddens=return_all_hiddens,
  File "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/ec2-user/.cache/torch/hub/pytorch_fairseq_master/fairseq/models/roberta/model.py", line 96, in forward
    x, extra = self.decoder(src_tokens, features_only, return_all_hiddens, **kwargs)
  File "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/ec2-user/.cache/torch/hub/pytorch_fairseq_master/fairseq/models/roberta/model.py", line 265, in forward
    x, extra = self.extract_features(src_tokens, return_all_hiddens)
  File "/home/ec2-user/.cache/torch/hub/pytorch_fairseq_master/fairseq/models/roberta/model.py", line 272, in extract_features
    src_tokens, last_state_only=not return_all_hiddens,
  File "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/ec2-user/.cache/torch/hub/pytorch_fairseq_master/fairseq/modules/transformer_sentence_encoder.py", line 210, in forward
    x, _ = layer(x, self_attn_padding_mask=padding_mask)
  File "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/ec2-user/.cache/torch/hub/pytorch_fairseq_master/fairseq/modules/transformer_sentence_encoder_layer.py", line 86, in forward
    x = self.activation_fn(self.fc1(x))
  File "/home/ec2-user/.cache/torch/hub/pytorch_fairseq_master/fairseq/modules/gelu.py", line 25, in gelu
    return x * 0.5 * (1.0 + torch.erf(x / math.sqrt(2.0)))
RuntimeError: [enforce fail at CPUAllocator.cpp:56] posix_memalign(&data, gAlignment, nbytes) == 0. 12 vs 0
```

i m assuming it has to do with preprocessing to batches. 

here's the code to batching the data:
```
class Batch(object):
    def _pad(self, data, pad_id, width=-1):
        if (width == -1):
            width = max(len(d) for d in data)
        rtn_data = [d + [pad_id] * (width - len(d)) for d in data]
        return rtn_data

    def __init__(self, data=None, device=None,  is_test=False):
        """Create a Batch from a list of examples."""
        if data is not None:
            self.batch_size = len(data)
            roberta = torch.hub.load('pytorch/fairseq', 'roberta.base', force_reload=True)
           
            pre_src = collate_tokens([x['src'] for x in data], pad_idx=1)
            
            pre_src_seq = [x['src'] for x in data]
            pre_tgt_seq = [x['tgt'] for x in data]
            # pad tokens and batch together
            # batch x sequence
            src_seq = collate_tokens(pre_src_seq, pad_idx=1)
            tgt_seq = collate_tokens(pre_tgt_seq, pad_idx=1)


            # batch x sequence -> batch x sequence x hidden tensor shape
            src = roberta.extract_features(pre_src)

            setattr(self, 'src', src.to(device))
            setattr(self, 'src_seq', src_seq.to(device))
            setattr(self, 'tgt_seq', tgt_seq.to(device))


            if (is_test):
                src_txt = [x['src_txt'] for x in data]
                setattr(self, 'src_txt', src_txt)
                tgt_txt = [x['tgt_txt'] for x in data]
                setattr(self, 'tgt_txt', tgt_txt)

    def __len__(self):
        return self.batch_size
```
I m using collate_tokens as well as extract_features. The question is do the provided functions work together like this or should I use a different approach? I m assuming the above code is the problem

**Comment by myleott** [comment]
I've never seen that particular error before, but I don't think your code is quite right, because you need to use `roberta.encode` to generate the proper token IDs. Please refer to the instructions here: https://github.com/pytorch/fairseq/tree/master/examples/roberta#batched-prediction

If you're training on GPU, then you should also call `roberta.cuda()` at some point.

**Comment by rush86999** [comment]
I will update this thread after making ```roberta.cuda()``` (thanks for the heads up) and possibly changing the batch_size. The values in the dictionary were saved using ```torch.load``` after being converted using ```roberta.encode``` on them.  Those values are token IDs pickled onto local ```.pt`` file and then loaded into batches. After searching for this problem there is a mention of possibly a large batch_size so I will let you know.  Thanks for your help

**Comment by rush86999** [comment]
I was able to update the code as per your suggestions and currently, I am training a model with batch_size = 8 with each row being a single sequence.  hopefully, I get something useful out of this. thanks for your patience. I believe the error was related to ```roberta.cuda()``` . I also encountered the error of maxing out memory limit. This led to a further halving of batch_size. Finally one other issue was changing the tensors to ```cuda``` device in the batch class mentioned above by ```seq = seq.to(device)```. It finally worked but training is slow. I will let it finish. Again thanks for the help and hopefully this info will help someone else as well.
